### PR TITLE
Add OpenAPI v3.0 Docs for API

### LINKED ISSUE #75: API Documentation
We are using django-rest-framework so it is a very low hanging fruit to exploit the API. In fact, some developers seem to be using it before any documentation has been made. This is due to how simple and easy it is to explore the robosats API. Will summarize here everything you need to know to use it.

## Short API documentation

Simply open your Tor Browser (http://robosats6tkf3eva7x2voqso3a5wcorsnw34jveyxfqi2fu7oyheasid.onion) and check the following endpoints for full details:
#### Order limits for every currency pair
```
/api/limits/
```
#### Historical exchange activity
```
/api/historical/
```

#### All Market Ticks
```
/api/ticks/
```

#### General exchange information
```
/api/info/
```

#### Exchange rates and premiums
```
/api/price/
```
For all national currencies for non-KYC BTC in the RoboSats market (this data has some limitations)

#### Every public order in the book
```
/api/book/?currency=0&type=2
```
Use the [currency number](https://github.com/Reckless-Satoshi/robosats/blob/main/frontend/static/assets/currencies.json) to query only public orders for one currency, use 0 for all. Types, use 0 for BUY orders, 1 for SELL orders, 2 for ALL.

#### New user or login
```
/api/user/?token=
```
Create a new user or login if there is an existing user. Token should satisfy minimum entropy constrains

#### Make an order
```
/api/make/
```

#### Orders
```
/api/order/?order_id=
```
Check order information. Use the POST method for updating the order. By updating the order when needed, a complete trade pipeline can be completed just by using this endpoint.

-----
The only thing that is not accessible trough those endpoints is the chatrooms.

### LINKED ISSUE #249: Authentication that does not require CSRF tokens
This is unfortunately required for LAN access on Robosats for Embassy

### PR DESCRIPTION
**Pull request by redphix**
This PR fixes #75 

This PR aims to document all the routes, responses and request objects as OpenAPI spec. You can browse the Redoc Documentation UI at `/api/`.

As suggestd by @ShatteredBunny, went with `drf-spectacular`. Easy to use and get's the job done. It's ability to use seriallizers already defined to generate the api spec is spectacular indeed.

After merging of the PR, we should have complete OpenAPI v3 documentation UI available for the current API at `/api` or `/api/v0/` on production.
Some more work in remaining, which needs more understanding and reading of what the `/order` route does carefully.

TODO:
- [x] Going through code and logic and carefully looking at all possibilities of responses and possible and documenting them.
- [x] `/order` route (Needs more reading of the code)
- [x] Write intro and include Robosats logo
-  ~Write how to do Auth (How to include `Authorization` header properly)~ (To be done after fix #249 )
<hr>

[TODO](https://github.com/Reckless-Satoshi/robosats/issues/75#issuecomment-1218289407) from issue #75 
> ### Backend
> - [x] Expose OpenAPI definition endpoint (JSON). If it is possible to do everything using Django restframework, that's best; otherwise e.g. drf-spectacular could be used
> - [x] Add UI replacing this issue, e.g. SwaggerUI
> - [x] Make sure the API definition is complete, in particular that all request and response types and parameters are available and correct. If possible this is auto-generated, but maybe metadata in form of annotations is necessary. Also add the descriptions from this issue

### CODE REVIEW

**Comment by Reckless-Satoshi** [comment]
Just checked around on `/api/` and this is very very top! Super useful explanations of all the endpoints and very useful examples. I am feeling the urge to build something over this API :yum: 

I realized all of the Django `decimalFields` (...numerical) are `strings`, unexpected to me.

File: api/views.py [hunk 1]
```diff
@@ -626,6 +764,152 @@ def get(self, request, format=None):
             context["bad_request"] = "User Generation with explicit token deprecated. Only token_sha256 allowed."
             return Response(context, status.HTTP_400_BAD_REQUEST)
 
+    @extend_schema(
```
Comment (Reckless-Satoshi) [inline api/views.py:767]: I wonder whether we can offload some lines of code to other files. Maybe we can create `schemas.py`  and import here somehow? It starts to get a bit worrisome to work on such a huge file that happens to be the most critical script on the platform :D

File: api/views.py [hunk 2]
```diff
@@ -1018,7 +1517,39 @@ def get(self, request):
 
         return Response(payload, status.HTTP_200_OK)
 
+
 class HistoricalView(ListAPIView):
+
+    @extend_schema(
+        summary='Get historical cxchange activity',
```
Comment (Reckless-Satoshi) [inline api/views.py:1524]: Small typo?

**Comment by redphix** [comment]
>  I am feeling the urge to build something over this API :yum: 

:smile: 

> I realized all of the Django decimalFields (...numerical) are strings, unexpected to me.

Yes, that was wired to me as well. But my guess is because of the constraints `max_digits`, `decimal_places` it makes it into strings so that it can enforce those constraints using regex? Not super convinced myself that this is the reason though and it does seem like a bug maybe. Didn't play around changing that model type though. But this should be easily fixable using serializers.

File: api/views.py (hunk 1 above)
Comment (redphix) [reply api/views.py:767]: Agreed. Good idea. Will split them into `schemas.py`

Comment (redphix) [reply api/views.py]: `schemas.py` seemed too generic and ambiguous. Went for `oas_schemas.py` instead. Implying that these schemas are only for the OpenAPI spec definition.

**Comment by redphix** [comment]
@Reckless-Satoshi It's taken longer than expected to complete this. Reading through all the logic flows and figuring out when some field is sent in the response was really time consuming. I tried my best to document it, but I may have missed some. In the API refactor we should consider making dedicated serializers for each type of response we expect - If that is too much (and there is only one way to know for sure, i.e by doing it) we can use jsonschema to it's fullest capabilities to express the models (eg Status.WFB model will be different from say Status.PAY model). Nonetheless, that's a task of it's own. 
As mentioned earlier, the new serializers or only for documenting purpose and are not used to do actual serialization. 

PS. will get to POST /order soon. Hopefully should not take as long as it took for GET /order
Also, lmk if required, I will reduce the scope of this PR and open another one, if you wish

**Comment by Reckless-Satoshi** [comment]
> @Reckless-Satoshi It's taken longer than expected to complete this. Reading through all the logic flows and figuring out when some field is sent in the response was really time consuming.

Hey @redphix , to be honest the work you are doing on this PR goes far and beyond what was expected for the task. You have had to dive very deep into the models and logics to understand and disentangle the full Order endpoint (I can only image how much work that took!). Amazing contribution! 

The API documentation page is looking gorgeous, only reviewing it is a task of its own :) I will be checking over it during next days and leave comments if necessary. All in all, the work you are doing here opens up a much more clear picture for devs and the ways to go for improving the API. A major revision of it is due.


> Also, lmk if required, I will reduce the scope of this PR and open another one, if you wish

I think we can keep stacking commits on this PR.

**Comment by redphix** [comment]
> Amazing contribution!

😄 Thank you!

> All in all, the work you are doing here opens up a much more clear picture for devs and the ways to go for improving the API. A major revision of it is due.

Yup. Gives a clearer picture right away on what has to be improved. I have a list of things, that I shall share soon.

**Comment by redphix** [comment]
Wew! That was a nice ride down the codebase! I think I have got a good understanding of the trade pipeline and it's inner workings in so detail now. I should be able to use this knowledge to write some docs as well soon!

One thing is clear. The API needs re-writing! hahaha. As mentioned in DMs @Reckless-Satoshi, I should have a issue out soon in which I lay out some problems and give, in my opinion, how we should solve it and we can discuss there. 

All the routes should be complete with openapi spec down to each field that the api sends, except for the POST /order route. I left it because I think it's going to be more wasted work because it's better to first move to v1 API and then do the second part of #75 (i.e using openapi to generate code that the FE can use). There are tons of improvements to be made as mentioned earlier. I have made sure to do the least wasteful work possible. All the explanations of the various fields can be carried over to v1 later on as well, and some serializers defined here can also be re-used in v1

@Reckless-Satoshi This is a lot to review, I understand, so take your time :)

**Comment by redphix** [comment]
Also the openapi yaml/json files can be obtained by:
YAML format:
```
docker exec -it django-dev python3 manage.py spectacular --file openapi.yaml
```
JSON format: 
```
docker exec -it django-dev python3 manage.py spectacular --format openapi-json --file openapi.json
```

File: api/views.py [hunk 3]
```diff
@@ -561,7 +573,7 @@ class UserView(APIView):
 
     def get(self, request, format=None):
```
Comment (Reckless-Satoshi) [inline api/views.py:574]: This GET to UserView has been deleted in main (it's the old way to generate users before we were using PGP). I guess this is the cause of the current merge conflict

File: api/oas_schemas.py [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+)
+
+EXP_MAKER_BOND_INVOICE = int(config("EXP_MAKER_BOND_INVOICE"))
+RETRY_TIME = int(config("RETRY_TIME"))
+PUBLIC_DURATION = 60 * 60 * int(config("DEFAULT_PUBLIC_ORDER_DURATION")) - 1
+ESCROW_DURATION = 60 * int(config("INVOICE_AND_ESCROW_DURATION"))
+BOND_SIZE = int(config("DEFAULT_BOND_SIZE"))
+
+
+class MakerViewSchema:
```
Comment (Reckless-Satoshi) [inline api/oas_schemas.py:19]: There is so much work and attention to detail here that is overwhelming. I say it is accurate to over 99% and the few things that are not fully accurate are not worth fixing. All in all I would suggest to merge as is: having this doc for the v0 API is a huge step forward.

**Review by Reckless-Satoshi — APPROVED** [review]
Hey @redphix ,

This is a really great contribution. I think we should merge as is. Descriptions are concise and accurate. There might be some items that are not fully accurate but I do not think improvement is needed.

As mentioned before I think the work here goes beyond what expected when the task was created. So let's double the planned 300K rewards. Please submit an invoice from a proxy node_id for 600K Sats.

Looking forward to your next contributions to improve the API. Would be great if we come up with a concrete action plan that can be split into several sub-tasks. :rocket:

**Comment by Reckless-Satoshi** [comment]
Can you take care of solving the merge conflict too?

File: api/serializers.py [hunk 5]
```diff
@@ -28,7 +57,421 @@ class Meta:
...
+        required=False,
+        help_text=f"Timestamp of when the contract was finalized (price and sats fixed)"
+    )
+    contract_total_time = serializers.FloatField(
+        required=False,
+        help_text="The time taken for the contract to complete (from taker taking the order to completion of order) in seconds"
+    )
+    routing_fee_sats = serializers.IntegerField(
+        required=False,
+        help_text="Sats payed by the exchange (currently const 0). Mining fee in case of on-chain swap payout"
```
[the remark this answers is not in the document]
Comment (redphix) [reply api/serializers.py:116]: ahh is it? I may have missed that I was in a `if swap` block. Will just remove that.

File: api/oas_schemas.py (hunk 4 above)
Comment (redphix) [reply api/oas_schemas.py:19]: hehe.. Appreciate you noticing :grin: . Tried my best to me accurate and detailed as possible.

**Comment by redphix** [comment]
> So let's double the planned 300K rewards.

Awesome! Much appreciated! :D

```
lnbc6m1p3nnjycpp53zxaupmfaxpzjlgzeewvj9m3wr7cvv3sg02lfs5hdsxmtugsjlksdqqcqzpgxqyd9uqsp5e5l5h2hgfkkfaf0lgeq33h2rghdr0xgr6v838k89gwgtsltkpjts9qyyssqp0afyn9cf5kysfwh0nj99nzf6etd6sl53h4run78pjwg5hmpnj44hf4r0yyd8rf79zflde708fxelk43lyatek42g2z9zcqquupeqaqpevvgzr
```

**Comment by redphix** [comment]
> Would be great if we come up with a concrete action plan that can be split into several sub-tasks. :rocket:

:+1: Will open an issue soon.

**Comment by Reckless-Satoshi** [comment]
> ```
> lnbc6m1p3nnjycpp53zxaupmfaxpzjlgzeewvj9m3wr7cvv3sg02lfs5hdsxmtugsjlksdqqcqzpgxqyd9uqsp5e5l5h2hgfkkfaf0lgeq33h2rghdr0xgr6v838k89gwgtsltkpjts9qyyssqp0afyn9cf5kysfwh0nj99nzf6etd6sl53h4run78pjwg5hmpnj44hf4r0yyd8rf79zflde708fxelk43lyatek42g2z9zcqquupeqaqpevvgzr
> ```
888dde0769e982297d02ce5cc9177170fd86323043d5f4c2976c0db5f11097ed

**Pull request merged**